### PR TITLE
[iter] Add hb_reduce

### PR DESCRIPTION
**Pull request by ebraminio**
These will be nice to have also:
* <s>Being able to have accumulator in a different type</s> (done)
* <s>hb_reduce_right (or hb_reduce_end?)</s> (not needed, we should have reverse instead)
* <s>Make it work somehow without initial value also</s> (possible but a respective default type also is needed) 

This alternatively can be named fold, accumulate, aggregate, compress, or inject [also](https://en.wikipedia.org/wiki/Fold_(higher-order_function)) and things like sum/min/max/average can be implemented upon it.

### CODE REVIEW

**Comment by behdad** [comment]
Thanks.  I have a long list of things to do with the iter stuff.  You can follow in https://github.com/harfbuzz/harfbuzz/issues/1648

**Comment by behdad** [comment]
Thanks.  Easy enough to do without initial value.  let's do that as well?  Just an overload.

**Comment by ebraminio** [comment]
> Easy enough to do without initial value.

Can't see how to do it without specifying a default type however.

File: src/hb-iter.hh [hunk 1]
```diff
@@ -344,6 +344,38 @@ static const struct
...
+};
+static const struct
+{
+  template <typename Redu, typename TValue = int> hb_reduce_t<Redu, TValue>
+  operator () (Redu&& r, TValue init_value = 0) const
+  { return hb_reduce_t<Redu, TValue> (r, init_value); }
+
+  template <typename Redu, typename TValue = int> hb_reduce_t<Redu&, TValue>
+  operator () (Redu *r, TValue init_value = 0) const
+  { return hb_reduce_t<Redu&, TValue> (*r, init_value); }
```
Comment (behdad) [inline src/hb-iter.hh]: I don't think we need a separate overload for pointer.  For now remove.  When we address that issue across the board, I think we just need to add a `hb_deref_pointer` to the existing implementation.

**Comment by behdad** [comment]
Humm, yeah no this is not good.  I think we might need to use two different types for the two cases.  I don't mind reverting back to the simple case you had.

Also, `reduce_right` isn't quite in scope for us.  That should be done with something like a hb_reverse() in between.  Or maybe only if iter is random-access.  Anyway, no need right now.

**Comment by behdad** [comment]
Thanks.  One more round and we'd be done: the accumulation type is not necess the same as initial type.  Accumulation type should be `decltype(hb_declval (Redu) (hb_declval (InitType), Iter:__item_type__))`.  Something like that.  Do you agree?

**Comment by behdad** [comment]
Eg. init might be integer, but acummulation type should be float if the reducer returns float.

**Comment by behdad** [comment]
> So what happens when the array is empty. Accumulation type and initial type should be in the same type but the iterator values can be in a different type, what the second test demonstrates.

You assign to accumulation type the initial value and let the type system figure out the conversion.

**Comment by ebraminio** [comment]
> You assign to accumulation type the initial value and let the type system figure out the conversion.

That won't do the thing I'm seeking in the second test IIUC.

**Comment by behdad** [comment]
> That won't do the thing I'm seeking in the second test IIUC.

How does it not?  If iterator is empty, you end up just returning the initial type.  I'm not sure what you are talking about.

**Comment by ebraminio** [comment]
Ok. Definitely makes sense. I am struggling on how to do it :)

**Comment by behdad** [comment]
I wrote it down for you ;).

I suggest this:
```
template <AccT = decltype (...)>
operator () (...)
```

**Comment by behdad** [comment]
Ie. accumulation type is not known until the operator(), since it depends on the iterator item type.

**Comment by ebraminio** [comment]
Done

**Comment by behdad** [comment]
Looks great.  Thanks :).

**Pull request merged**